app/src/openxr/cpp/OpenXRGestureManager.cpp

```diff
@@ -0,0 +1,112 @@
+//
+// Created by sergio on 05/07/23.
+//
+
+#include <cstring>
+#include "OpenXRGestureManager.h"
+
+namespace crow {
+
+// This threshold is used to detect when palm is facing the head,
+// and enable the left hand action gesture. The higher the threshold
+// the more aligned objects should be to be considered facing each other.
+// 0.7 is generally accepted as good for objects facing each other.
+const float kPalmHeadThreshold = 0.7;
+
+bool
+OpenXRGestureManager::palmFacesHead(const vrb::Matrix &palm, const vrb::Matrix &head) const {
+    // For the hand we take the Y axis because that corresponds to head's Z axis when
+    // the hand is in upright position facing head (the gesture we want to detect).
+#ifdef PICOXR
+    // Axis are inverted in Pico
+    auto vectorPalm = palm.MultiplyDirection({0, 0, -1});
+#else
+    auto vectorPalm = palm.MultiplyDirection({0, 1, 0});
+#endif
+    auto vectorHead = head.MultiplyDirection({0, 0, -1});
+    return vectorPalm.Dot(vectorHead) > kPalmHeadThreshold;
+}
+
+void
+OpenXRGestureManagerFBHandTrackingAim::populateNextStructureIfNeeded(XrHandJointLocationsEXT& handJointLocations) {
+    memset(&mFBAimState, 0, sizeof(XrHandTrackingAimStateFB));
+    mFBAimState = { XR_TYPE_HAND_TRACKING_AIM_STATE_FB, handJointLocations.next, 0  };
+    handJointLocations.next = &mFBAimState;
+}
+
+bool
+OpenXRGestureManagerFBHandTrackingAim::hasAim() const {
+    return mFBAimState.status & XR_HAND_TRACKING_AIM_VALID_BIT_FB;
+}
+
+XrPosef
+OpenXRGestureManagerFBHandTrackingAim::aimPose(const XrTime predictedDisplayTime, const OpenXRHandFlags handeness, const vrb::Matrix& head) const {
+    ASSERT(hasAim());
+    return mFBAimState.aimPose;
+}
+
+bool
+OpenXRGestureManagerFBHandTrackingAim::systemGestureDetected(const vrb::Matrix& palm, const vrb::Matrix& head) const {
+#ifdef PICOXR
+    // Pico does not correctly implement the SYSTEM_GESTURE_BIT_FB flags.
+    return palmFacesHead(palm, head) && !hasAim();
+#else
+    return mFBAimState.status & XR_HAND_TRACKING_AIM_SYSTEM_GESTURE_BIT_FB;
+#endif
+}
+
+OpenXRGestureManagerHandJoints::OpenXRGestureManagerHandJoints(HandJointsArray& handJoints)
+    : mHandJoints(handJoints)
+    , mOneEuroFilterPosition(std::make_unique<OneEuroFilterVector>(0.25, 0.1, 1)) {
+}
+
+bool
+OpenXRGestureManagerHandJoints::hasAim() const {
+    return IsHandJointPositionValid(XR_HAND_JOINT_MIDDLE_PROXIMAL_EXT, mHandJoints);
+}
+
+XrPosef
+OpenXRGestureManagerHandJoints::aimPose(const XrTime predictedDisplayTime, const OpenXRHandFlags handeness, const vrb::Matrix& head) const {
+    ASSERT(hasAim());
+    auto lookAt = [](const vrb::Vector& sourcePoint, const vrb::Vector& destPoint) -> vrb::Quaternion {
```

Comment:
Is this lambda really needed ? it seems it's used only once in the context of this function.

Reply:
This is just moving code. I preferred not to change anything from the original code.

That said, yeah I think it's good to have that code in a lambda as it nicely removes the complexity from the main execution path which does only call a couple of functions.